proposals/2022-09-14-kubernetes-versioning.md

```diff
@@ -0,0 +1,208 @@
+
+## <a name='Background'></a>Background
+
+It is likely behaviors in a shim may change over time as new features are introduced or as existing defects are resolved. Pod workloads that target a version N of a shim may not be able to execute on version N+1. This proposal describes a solution to ensure Pod workloads targeting a specific version of a shim will always be scheduled on a node that contains a safe version of a shim for the workload.
+
+## <a name='ProblemStatement'></a>Problem Statement
+The containerd shims in this repository are installed on Kubernetes nodes. The containerd config on each Kubernetes node has a handler registered for each shim, a handler name mapped to binary on the PATH. A RuntimeClass is registered in Kubernetes describing the containerd handler needed to run a Pod workload, as well as label selectors to inform the Kubernetes scheduler to only schedule workloads on particular nodes.
+
+1. A shim binary is installed on a Kubernetes node at `/bin/containerd-shim-spin-v1`.
+2. The following shows the containerd config.toml that maps the runtime type to the shim binary from step 1.
+      ```toml
+        [plugins.cri.containerd.runtimes.spin]
+          runtime_type = "io.containerd.spin.v1"
+      ```
+3. A runtime class that contains a handler that matches the "spin" config runtime name from step 2.
+      ```yaml
+      apiVersion: node.k8s.io/v1
+      kind: RuntimeClass
+      metadata:
+        name: wasmtime-spin
+      handler: spin
+      ```
+    
+    **NOTE: if no scheduling configuration is specified, the Kubernetes scheduler expects every node is able to run the workload.**
+
+4. A Wasm workload with the specified runtime class name matching the "wasmtime-spin" runtime class from step 3.
+      ```yaml
+      apiVersion: apps/v1
+      kind: Deployment
+      metadata:
+        name: wasm-spin
+      spec:
+        replicas: 1
+        selector:
+          matchLabels:
+            app: wasm-spin
+        template:
+          metadata:
+            labels:
+              app: wasm-spin
+          spec:
+            runtimeClassName: wasmtime-spin
+            containers:
+            - name: spin-hello
+              image: ghcr.io/deislabs/containerd-wasm-shims/examples/spin-rust-hello:latest
+              command: ["/"]
+      ```
+
+The above configuration will work well if there are no breaking changes introduced to the shim binary or if no new features are added that a given workload depends upon. However, this scenario is unlikely given the speed of innovation in this space.
+
+To safe introduce new nodes and new versions of shims, we must have a versioning scheme for both the shim binary and the RuntimeClasses.
+
+### <a name='Goals'></a>Goals
+1) A user should be able to add new nodes to the cluster with new versions of the shims and not disturb existing workloads.
+2) A user should be able to target a major, major + minor, or major+minor+patch semantic version of a shim.
```

Comment:
Do you like the idea that the shim binary version follows the underlying wasm runtime version? For example, if the spin shim is using spin `v0.5.1`, then the shim's binary version is also `v0.5.1`. 

There is a downside though. We can't use the main branch of spin anymore, because main branch release of spin is called `canary` release. If we follow the rule above, then the shim's binary version is also `canary`. But this makes the version no longer following semantic versioning...

Reply:
Hmm... The version of our shim doesn't need to follow spin's versioning. In fact, I can imagine many times where our version should increment faster than Spin's due to changes in the way the shim executes.